# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: editor/libeditor/EditorUtils.cpp
# Commit: a43e95a534eb
# Full Hash: a43e95a534eb5c13883cc2a61fb64c15fe29e674
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2022-10-05 03:40:21
# Regressor Bug: 1792654
# File Overlap Count: 1
# Description:
#   Bug 1792654 - Make `CreateNodeResultBase`, `MoveNodeResult`, `SplitNodeResult` and `SplitRangeOffResult` handle caret point with a common base class r=m_kato
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D158335
# ==============================================================================

diff -r 18196c5fab62 -r a43e95a534eb editor/libeditor/EditorUtils.cpp
--- a/editor/libeditor/EditorUtils.cpp	Tue Oct 04 21:13:53 2022 +0000
+++ b/editor/libeditor/EditorUtils.cpp	Tue Oct 04 21:36:57 2022 +0000
@@ -448,15 +448,10 @@
 }
 
 /******************************************************************************
- * mozilla::CreateNodeResultBase
+ * mozilla::CaretPoint
  *****************************************************************************/
 
-NS_INSTANTIATE_CREATE_NODE_RESULT_CONST_METHOD(
-    nsresult, SuggestCaretPointTo, const EditorBase& aEditorBase,
-    const SuggestCaretOptions& aOptions)
-
-template <typename NodeType>
-nsresult CreateNodeResultBase<NodeType>::SuggestCaretPointTo(
+nsresult CaretPoint::SuggestCaretPointTo(
     const EditorBase& aEditorBase, const SuggestCaretOptions& aOptions) const {
   mHandledCaretPoint = true;
   if (!mCaretPoint.IsSet()) {
@@ -481,15 +476,26 @@
              : rv;
 }
 
-NS_INSTANTIATE_CREATE_NODE_RESULT_METHOD(bool, MoveCaretPointTo,
-                                         EditorDOMPoint& aPointToPutCaret,
-                                         const EditorBase& aEditorBase,
-                                         const SuggestCaretOptions& aOptions)
+bool CaretPoint::CopyCaretPointTo(EditorDOMPoint& aPointToPutCaret,
+                                  const EditorBase& aEditorBase,
+                                  const SuggestCaretOptions& aOptions) const {
+  MOZ_ASSERT(!aOptions.contains(SuggestCaret::AndIgnoreTrivialError));
+  mHandledCaretPoint = true;
+  if (aOptions.contains(SuggestCaret::OnlyIfHasSuggestion) &&
+      !mCaretPoint.IsSet()) {
+    return false;
+  }
+  if (aOptions.contains(SuggestCaret::OnlyIfTransactionsAllowedToDoIt) &&
+      !aEditorBase.AllowsTransactionsToChangeSelection()) {
+    return false;
+  }
+  aPointToPutCaret = mCaretPoint;
+  return true;
+}
 
-template <typename NodeType>
-bool CreateNodeResultBase<NodeType>::MoveCaretPointTo(
-    EditorDOMPoint& aPointToPutCaret, const EditorBase& aEditorBase,
-    const SuggestCaretOptions& aOptions) {
+bool CaretPoint::MoveCaretPointTo(EditorDOMPoint& aPointToPutCaret,
+                                  const EditorBase& aEditorBase,
+                                  const SuggestCaretOptions& aOptions) {
   MOZ_ASSERT(!aOptions.contains(SuggestCaret::AndIgnoreTrivialError));
   mHandledCaretPoint = true;
   if (aOptions.contains(SuggestCaret::OnlyIfHasSuggestion) &&